parser/src/lib.rs

```diff
@@ -134,32 +139,91 @@ type ErrorEnv<'err, 'input> = &'err mut Errors<LalrpopError<'input>>;
 
 pub type ParseErrors = Errors<Spanned<Error, BytePos>>;
 
+struct Lexer<'input, Tokens> {
+    tokens: Tokens,
+    layout: LayoutContext<'input>,
+}
+
+impl<'input, Tokens> Lexer<'input, Tokens>
+    where Tokens: Iterator<Item = Result<SpannedToken<'input>, TokenizeError>>,
+{
+    fn new(tokens: Tokens) -> Lexer<'input, Tokens> {
+        Lexer {
+            tokens: tokens,
+            layout: LayoutContext::new(),
+        }
+    }
+
+    fn next_token(&mut self) -> Option<Result<SpannedToken<'input>, Error>> {
+        match self.tokens.next() {
```

Comment:
Updated in https://github.com/gluon-lang/gluon/pull/103/commits/7586a0ff4b07be8c5c17134cfd586e072f79afac to be clearer